### ISSUE TITLE
TimeoutError: There are still unapplied migrations after 60 seconds

### ISSUE THREAD

**Issue by nike4122** [airflow-helm/charts#213]
Hi,

I have done the setup of airflow 2.0.2 using helm chart in AKS. (https://artifacthub.io/packages/helm/airflow-helm/airflow).
I have configured external postgresql with airflow, but after the connection of external DB pods are not coming up.

NAME                                                                READY   STATUS                  
pod/apache-airflow-scheduler-5c9f99cbf7-d798b   0/1     Init:1/2              
pod/apache-airflow-scheduler-7d4859d696-r4ljg   0/1     Init:1/2           
pod/apache-airflow-web-798656c9ff-75fhf         0/1     Init:CrashLoopBackOff   
pod/apache-airflow-web-7d7d9ddf8-mc542          0/1     Init:CrashLoopBackOff   

I am getting below logs from wait-for-db-migrations container,

[2021-05-27 16:43:39,574] {migration.py:163} INFO - Context impl PostgresqlImpl.
[2021-05-27 16:43:39,575] {migration.py:170} INFO - Will assume transactional DDL.
[2021-05-27 16:43:43,947] {providers_manager.py:299} WARNING - Exception when importing 'airflow.providers.microsoft.azure.hooks.wasb.WasbHook' from 'apache-airflow-providers-microsoft-azure' package: No module named 'azure.storage.blob'
[2021-05-27 16:43:44,566] {providers_manager.py:299} WARNING - Exception when importing 'airflow.providers.microsoft.azure.hooks.wasb.WasbHook' from 'apache-airflow-providers-microsoft-azure' package: No module named 'azure.storage.blob'
[2021-05-27 16:43:45,655] {db.py:601} INFO - Waiting for migrations... 1 second(s)
[2021-05-27 16:43:46,661] {db.py:601} INFO - Waiting for migrations... 2 second(s)
[2021-05-27 16:43:47,668] {db.py:601} INFO - Waiting for migrations... 3 second(s)
[2021-05-27 16:43:48,674] {db.py:601} INFO - Waiting for migrations... 4 second(s)
[2021-05-27 16:43:49,680] {db.py:601} INFO - Waiting for migrations... 5 second(s)
[2021-05-27 16:43:50,688] {db.py:601} INFO - Waiting for migrations... 6 second(s)
[2021-05-27 16:43:51,697] {db.py:601} INFO - Waiting for migrations... 7 second(s)
[2021-05-27 16:43:52,703] {db.py:601} INFO - Waiting for migrations... 8 second(s)
[2021-05-27 16:43:53,708] {db.py:601} INFO - Waiting for migrations... 9 second(s)
[2021-05-27 16:43:54,714] {db.py:601} INFO - Waiting for migrations... 10 second(s)
[2021-05-27 16:43:55,722] {db.py:601} INFO - Waiting for migrations... 11 second(s)
[2021-05-27 16:43:56,730] {db.py:601} INFO - Waiting for migrations... 12 second(s)
[2021-05-27 16:43:57,736] {db.py:601} INFO - Waiting for migrations... 13 second(s)
[2021-05-27 16:43:58,742] {db.py:601} INFO - Waiting for migrations... 14 second(s)
[2021-05-27 16:43:59,747] {db.py:601} INFO - Waiting for migrations... 15 second(s)
[2021-05-27 16:44:00,752] {db.py:601} INFO - Waiting for migrations... 16 second(s)
[2021-05-27 16:44:01,758] {db.py:601} INFO - Waiting for migrations... 17 second(s)
[2021-05-27 16:44:02,763] {db.py:601} INFO - Waiting for migrations... 18 second(s)
[2021-05-27 16:44:03,768] {db.py:601} INFO - Waiting for migrations... 19 second(s)
[2021-05-27 16:44:04,773] {db.py:601} INFO - Waiting for migrations... 20 second(s)
[2021-05-27 16:44:05,785] {db.py:601} INFO - Waiting for migrations... 21 second(s)
[2021-05-27 16:44:06,800] {db.py:601} INFO - Waiting for migrations... 22 second(s)
[2021-05-27 16:44:07,807] {db.py:601} INFO - Waiting for migrations... 23 second(s)
[2021-05-27 16:44:08,812] {db.py:601} INFO - Waiting for migrations... 24 second(s)
[2021-05-27 16:44:09,818] {db.py:601} INFO - Waiting for migrations... 25 second(s)
[2021-05-27 16:44:10,824] {db.py:601} INFO - Waiting for migrations... 26 second(s)
[2021-05-27 16:44:11,829] {db.py:601} INFO - Waiting for migrations... 27 second(s)
[2021-05-27 16:44:12,834] {db.py:601} INFO - Waiting for migrations... 28 second(s)
[2021-05-27 16:44:13,838] {db.py:601} INFO - Waiting for migrations... 29 second(s)
[2021-05-27 16:44:14,844] {db.py:601} INFO - Waiting for migrations... 30 second(s)
[2021-05-27 16:44:15,849] {db.py:601} INFO - Waiting for migrations... 31 second(s)
[2021-05-27 16:44:16,856] {db.py:601} INFO - Waiting for migrations... 32 second(s)
[2021-05-27 16:44:17,860] {db.py:601} INFO - Waiting for migrations... 33 second(s)
[2021-05-27 16:44:18,869] {db.py:601} INFO - Waiting for migrations... 34 second(s)
[2021-05-27 16:44:19,898] {db.py:601} INFO - Waiting for migrations... 35 second(s)
[2021-05-27 16:44:20,909] {db.py:601} INFO - Waiting for migrations... 36 second(s)
[2021-05-27 16:44:21,915] {db.py:601} INFO - Waiting for migrations... 37 second(s)
[2021-05-27 16:44:22,920] {db.py:601} INFO - Waiting for migrations... 38 second(s)
[2021-05-27 16:44:23,936] {db.py:601} INFO - Waiting for migrations... 39 second(s)
[2021-05-27 16:44:24,942] {db.py:601} INFO - Waiting for migrations... 40 second(s)
[2021-05-27 16:44:25,947] {db.py:601} INFO - Waiting for migrations... 41 second(s)
[2021-05-27 16:44:26,952] {db.py:601} INFO - Waiting for migrations... 42 second(s)
[2021-05-27 16:44:27,956] {db.py:601} INFO - Waiting for migrations... 43 second(s)
[2021-05-27 16:44:28,961] {db.py:601} INFO - Waiting for migrations... 44 second(s)
[2021-05-27 16:44:29,967] {db.py:601} INFO - Waiting for migrations... 45 second(s)
[2021-05-27 16:44:30,972] {db.py:601} INFO - Waiting for migrations... 46 second(s)
[2021-05-27 16:44:31,977] {db.py:601} INFO - Waiting for migrations... 47 second(s)
[2021-05-27 16:44:32,982] {db.py:601} INFO - Waiting for migrations... 48 second(s)
[2021-05-27 16:44:33,987] {db.py:601} INFO - Waiting for migrations... 49 second(s)
[2021-05-27 16:44:34,992] {db.py:601} INFO - Waiting for migrations... 50 second(s)
[2021-05-27 16:44:35,999] {db.py:601} INFO - Waiting for migrations... 51 second(s)
[2021-05-27 16:44:37,004] {db.py:601} INFO - Waiting for migrations... 52 second(s)
[2021-05-27 16:44:38,009] {db.py:601} INFO - Waiting for migrations... 53 second(s)
[2021-05-27 16:44:39,021] {db.py:601} INFO - Waiting for migrations... 54 second(s)
[2021-05-27 16:44:40,027] {db.py:601} INFO - Waiting for migrations... 55 second(s)
[2021-05-27 16:44:41,032] {db.py:601} INFO - Waiting for migrations... 56 second(s)
[2021-05-27 16:44:42,038] {db.py:601} INFO - Waiting for migrations... 57 second(s)
[2021-05-27 16:44:43,043] {db.py:601} INFO - Waiting for migrations... 58 second(s)
[2021-05-27 16:44:44,048] {db.py:601} INFO - Waiting for migrations... 59 second(s)
[2021-05-27 16:44:45,052] {db.py:601} INFO - Waiting for migrations... 60 second(s)
Traceback (most recent call last):
  File "/home/airflow/.local/bin/airflow", line 8, in <module>
    sys.exit(main())
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/__main__.py", line 40, in main
    args.func(args)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/cli/cli_parser.py", line 48, in command
    return func(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/cli/commands/db_command.py", line 54, in check_migrations
    db.check_migrations(timeout=args.migration_wait_timeout)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/utils/db.py", line 598, in check_migrations
    raise TimeoutError(f"There are still unapplied migrations after {ticker} seconds.")
TimeoutError: There are still unapplied migrations after 60 seconds.

**Comment by thesuperzapper** [comment]
@nike4122 Is the post-install `XX-upgrade-db` job finishing successfully? 
You check the status of the `job` resources with `kubectl get jobs --namespace XX`.

Otherwise, its possible that your DB is really slow (and the migration is taking a while) so we might need to increase the [timeout passed](https://github.com/airflow-helm/charts/blob/airflow-8.1.3/charts/airflow/templates/_helpers/pods.tpl#L55) to `airflow db upgrade -t XX`.

**Comment by AndreaGiardini** [comment]
Same issue here. Chart version 8.1.3. default values.yaml file

Looks like the job is never created:
```
➜ k get jobs -n airflow
No resources found in airflow namespace.
```

**Comment by AndreaGiardini** [comment]
In my case I had to set `helmWait: true`

https://github.com/airflow-helm/charts/blob/main/charts/airflow/values.yaml#L3

**Comment by nike4122** [comment]
@thesuperzapper  Somehow this issue resolved after I removed existing DB then connected to a fresh airflow setup.
But now I am facing a different issue, when I trigger a DAG job worker pod is failed to start and throwing an error

$ kubectl get po -n airflow
NAME                                                   READY   STATUS    RESTARTS   AGE
apache-airflow-scheduler-5c9f99cbf7-457pb              1/1     Running   103        16h
apache-airflow-web-7d7d9ddf8-8hbpf                     1/1     Running   3          16h
helloworldhellotask.757795a7562d4ed5b6feb5821d3fa0d5   0/1     Error     0          10h

$ kubectl logs helloworldhellotask.757795a7562d4ed5b6feb5821d3fa0d5 -n airflow
**/entrypoint: line 109: BACKEND: unbound variable**


I am using 
CHART_VERSION=8.1.3
KubernatesExecutor
Airflow - 2.0.2

**Comment by thesuperzapper** [comment]
@nike4122 what are your values?

**Comment by nike4122** [comment]
@thesuperzapper This issue has been resolved after we put **'%40'** instead of **@** in the username,

## the user of the external database
  ##
  user: airflow_user%40postgres

Thank you for your support!

**Comment by cocampbe** [comment]
I am facing the same issue after updating from airflow image 2.0.1-py3.8 to 2.1.2-py3.8. 

> [2021-07-20 12:19:18,815] {db.py:619} INFO - Waiting for migrations... 58 second(s)
[2021-07-20 12:19:19,818] {db.py:619} INFO - Waiting for migrations... 59 second(s)
[2021-07-20 12:19:20,821] {db.py:619} INFO - Waiting for migrations... 60 second(s)
Traceback (most recent call last):
  File "/home/airflow/.local/bin/airflow", line 8, in <module>
    sys.exit(main())
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/__main__.py", line 40, in main
    args.func(args)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/cli/cli_parser.py", line 48, in command
    return func(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/cli/commands/db_command.py", line 54, in check_migrations
    db.check_migrations(timeout=args.migration_wait_timeout)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/db.py", line 613, in check_migrations
    raise TimeoutError(
TimeoutError: There are still unapplied migrations after 60 seconds. Migration Head(s) in DB: {'82b7c48c147f'} | Migration Head(s) in Source Code: {'a13f7613ad25'}

There is no job listed in the namespace for the db upgrade. I am using an external postgres db, not the one baked into the airflow image.

**Comment by cocampbe** [comment]
I check the check-db init container and it shows a successful connection to the postgres db.

**Comment by cocampbe** [comment]
After trying to figure this out, it finally resolved. I use fluxcd for helm installs. I did a rollback and at some point it successfully ran the db migration. I then made an update, rollback, etc. And after the last update it was successful. I also tried an update to chart 8.4.1, but helm barfed on not being able to update the clusterip spec in one of the deployments, or sts'. I revertedf to 8.0.9 since it has been successful in the past. SMH.

**Comment by thesuperzapper** [comment]
@cocampbe I am planning to remove the `upgrade-db` Job altogether, and replace it with a Deployment, the Job has caused many issues, see issue https://github.com/airflow-helm/charts/issues/317 for more the proposal.

**Comment by thesuperzapper** [comment]
@nike4122, I never considered a special character in the db username, but I have raised an issue https://github.com/airflow-helm/charts/issues/318, to automatically encode it for users.

But be warned, you will need to revert it to its un-encoded form in `externalDatabase.user` once this update happens (I will put a warning in the changelog).

**Comment by thesuperzapper** [comment]
@cocampbe BTW, the unpatchable issue with the `upgrade-db` Job is [fixed in `8.5.0`](https://github.com/airflow-helm/charts/blob/main/charts/airflow/CHANGELOG.md#850---2021-08-19), as we have replaced the (unpatchable) Job with a (patchable) Deployment called `db-migrations`.

I am interested to see if your other issues are resolved.

_PS: it's also worth looking into `8.5.0` for PgBouncer_

**Comment by cocampbe** [comment]
> @cocampbe BTW, the unpatchable issue with the `upgrade-db` Job is [fixed in `8.5.0`](https://github.com/airflow-helm/charts/blob/main/charts/airflow/CHANGELOG.md#850---2021-08-19), as we have replaced the (unpatchable) Job with a (patchable) Deployment called `db-migrations`.
> 
> I am interested to see if your other issues are resolved.
> 
> _PS: it's also worth looking into `8.5.0` for PgBouncer_

I finally got to use 8.5.1 today. And I upgraded to airflow 2.2.2. The update went well. I did not enable pgbouncer, but will look into that later. The only issue I had was that some data in one of the  tables apparently had bad data. I  think that is an airflow feature, not the helm chart. I just wanted to get back to you on this. Thanks for your help.
